pkg/scheduler/plugins/coescheduling/core/core.go

```diff
@@ -0,0 +1,618 @@
+	pgMgr := &PodGroupManager{
+		pgClient:             pgClient,
+		snapshotSharedLister: snapshotSharedLister,
+		pgLister:             pgInformer.Lister(),
+		podLister:            podInformer.Lister(),
+	}
+	gangCache := NewGangCache(args, podInformer.Lister(), pgInformer.Lister())
+	pgMgr.cache = gangCache
+	return pgMgr
+}
+
+func (pgMgr *PodGroupManager) OnPodAdd(obj interface{}) {
+	pgMgr.cache.onPodAdd(obj)
+}
+
+func (pgMgr *PodGroupManager) OnPodDelete(obj interface{}) {
+	pgMgr.cache.onPodDelete(obj)
+}
+
+func (pgMgr *PodGroupManager) OnPodGroupAdd(obj interface{}) {
+	pgMgr.cache.onPodGroupAdd(obj)
+}
+
+func (pgMgr *PodGroupManager) OnPodGroupDelete(obj interface{}) {
+	pgMgr.cache.onPodGroupDelete(obj)
+}
+
+func (pgMgr *PodGroupManager) Recover() {
+	pgMgr.cache.RecoverGangCache()
+}
+
+// ActivateSiblings stashes the pods belonging to the same PodGroup of the given pod
+// in the given state, with a reserved key "kubernetes.io/pods-to-activate".
+func (pgMgr *PodGroupManager) ActivateSiblings(pod *corev1.Pod, state *framework.CycleState) {
+	gangName := GetGangNameByPod(pod)
+	if gangName == "" {
+		return
+	}
+	gangId := GetNamespaceSplicingName(pod.Namespace,
+		gangName)
+	pods, err := pgMgr.podLister.Pods(pod.Namespace).List(labels.NewSelector())
+	if err != nil {
+		klog.Errorf("ActivateSiblings Failed to list pods belong to a Gang: %v", gangId)
+		return
+	}
+	toActivePods := make([]*corev1.Pod, 0)
+	for i := range pods {
+		gangIdTmp := GetNamespaceSplicingName(pods[i].Namespace,
+			GetGangNameByPod(pods[i]))
+		if gangIdTmp == gangId {
+			if pods[i].UID == pod.UID {
+				continue
+			}
+			toActivePods = append(toActivePods, pods[i])
+		}
+	}
+
+	if len(toActivePods) != 0 {
+		if c, err := state.Read(framework.PodsToActivateKey); err == nil {
+			if s, ok := c.(*framework.PodsToActivate); ok {
+				s.Lock()
+				for _, pod := range toActivePods {
+					namespacedName := GetNamespaceSplicingName(pod.Namespace, pod.Name)
+					s.Map[namespacedName] = pod
+				}
+				s.Unlock()
+			}
+		}
+	}
+}
+
+// PreFilter
+// if non-strict-mode, we only do step1 and step2:
+// i.Check whether childes in Gang has met the requirements of minimum number under each Gang, and reject the pod if negative.
+// ii.Check whether the Gang has been timeout(check the pod's annotation,later introduced at Permit section) or is inited, and reject the pod if positive.
+// iii.Check whether the Gang has met the scheduleCycleValid check, and reject the pod if negative.
+// iv.Try update scheduleCycle, scheduleCycleValid, childrenScheduleRoundMap as mentioned above.
+func (pgMgr *PodGroupManager) PreFilter(ctx context.Context, pod *corev1.Pod) error {
+	gangName := GetGangNameByPod(pod)
+	if gangName == "" {
+		return nil
+	}
+	gangId := GetNamespaceSplicingName(pod.Namespace,
+		gangName)
+	gang := pgMgr.cache.getGangFromCache(gangId)
```

Comment:
all get gang process is same, please packge to a new func.

Reply:
done